### PR TITLE
Wheel speed function optimized

### PR DESCRIPTION
**Pull request by riyan-noaman**
# TASK TITLE

Lowering use of repetitive tasks and variables in order to create efficiency

## Changes

majority of wheel speed variables omitted and replaced with arrays
using for loops rather then high numbers of if else statements

## Deployment

- Bullet list of changes to deployment 

## Proof of functionality (screenshot, logs, etc)

yet to be tested 

## Additional Notes

Code will not work in current condition, missing checks for TIM3 channels, will be added later tdy then tested

## Checklist

- [x] Your code builds clean without any errors or warnings
- [ ] Code has been formatted (clang-format, cmake-format, black, prettier, rustfmt, etc)
- [ ] Where possible, unit tests have been added
- [ ] Code is well-commented & understood

### CODE REVIEW

**Review by Nomalah — CHANGES_REQUESTED** [review]
Go through, re-read the style guide. Put the results of the process into CTRL_Data. Also, remove comments which are stupid and self describing. Ex. "Loop from 0 to NUM_WHEELS"

File: Core/Inc/control.h [hunk 1]
```diff
@@ -23,6 +26,8 @@
 #define RTD_TRACTIVE_VOLTAGE_OFF 200
 // Turn off flag when tractive voltage > 450
 #define RTD_TRACTIVE_VOLTAGE_ON 4500
+//number of teeth on wheel hub
+#define numTeeth 32;
```
Comment (Nomalah) [inline Core/Inc/control.h:30]:
```suggestion
#define NUM_TEETH 32;
```
Remember the style guide

File: Core/Src/control.c [hunk 2]
```diff
@@ -8,68 +8,36 @@
...
-		LEDCtrl();
-		osDelayUntil(tick += CTRL_PERIOD);
-	}
+void startControlTask(){
+    uint32_t tick = osKernelGetTickCount();
+    while(1){
+        for(int i = 0; i < 4; i++){
+            GRCprintf("Frequency %d= %d\n\r", (i+1), Ctrl_Data.wheel_freq[i]);
+        }
+
```
Comment (Nomalah) [inline Core/Src/control.c:30]: Remove this debug printing or replace it with a TRACE_PRINT
```suggestion
```

File: Core/Src/control.c [hunk 3]
```diff
@@ -89,39 +57,28 @@ void startControlTask(){
...
-		if(TIM4_OVC >= 2){
-			TIM4_freq = 0;
-		}
-	}
+
+
+const TIM_HandleTypeDef* wheel_to_timer_mapping[NUM_WHEELS] = {&htim2, &htim3, &htim3, &htim4}; // Note: TIM3 is repeated for both channels CH1 and CH2
+
+// Function to manage overflow conditions for each wheel's timer
+void ManageTimerOverflow(TIM_HandleTypeDef * htim) {
```
Comment (Nomalah) [inline Core/Src/control.c:66]:
```suggestion
void manageWheelSpeedTimerOverflow(const TIM_HandleTypeDef * htim) {
```
Watch the style guide, and if you name it better, you don't need the comment describing what the function does.

File: Core/Src/control.c [hunk 4]
```diff
@@ -89,39 +57,28 @@ void startControlTask(){
...
+            if (tim_ovc[i] >= 2) {
+            	Ctrl_Data.wheel_freq[i] = 0;
+            }
+        }
+    }
 }
 
+
 /*
  * Called when an input capture event occurs
```
Comment (Nomalah) [inline Core/Src/control.c:83]: This comment block should be moved elsewhere & updated to represent the current status of the code. It should not describe exactly what the code does. Comments should not describe what something does (unless it's not obvious), comments should describe *why* something is done.

File: Core/Src/control.c [hunk 5]
```diff
@@ -148,135 +105,36 @@ void OverflowCheck(TIM_HandleTypeDef * htim){
...
+                // Calculate RPM and store in Ctrl_Data.wheel_rpm array
+                Ctrl_Data.wheel_rpm[i] = (Ctrl_Data.wheel_freq[i] * 60) / numTeeth; // assuming numTeeth is defined elsewhere
+            }
+            // Reset overflow counter
+            tim_ovc[i] = 0;
+            // Toggle state (0 to 1, or 1 to 0)
+            tim_state[i] = !tim_state[i];
+        }
+    }
 }
```
Comment (Nomalah) [inline Core/Src/control.c:136]: Remove like 99% of these comments. "store timer value within rising edge variable" is a pointless comments.
Don't comment what you changed, that's what git is for.

**Review by Nomalah — APPROVED** [review]
Style guide and comment points again.

**Review by Nomalah — CHANGES_REQUESTED** [review]
I actually wanted to request changes

**Review by Nomalah — CHANGES_REQUESTED** [review]
I actually wanted to request changes

File: Core/Src/control.c [hunk 6]
```diff
@@ -70,231 +40,101 @@ void startControlTask(){
...
+    }
+	// Calculate time interval between two rising edges at the same sensor
+	uint32_t ticks_TIM = (tim_rising[wheel_index][tim_state[wheel_index]] + (tim_ovc[wheel_index] * htim->Init.Period)) - tim_rising[wheel_index][!tim_state[wheel_index]];
+	// Calculate frequency and RPM if ticks is not zero and overflow counter is less than 2
+	if (ticks_TIM != 0 && tim_ovc[wheel_index] < 2) {
+		// Calculate frequency
+		Ctrl_Data.wheel_freq[wheel_index] = (uint32_t) (96000000UL / (htim->Init.Prescaler + 1) / ticks_TIM);
+		// Calculate RPM and store in Ctrl_Data.wheel_rpm array
+		Ctrl_Data.wheel_rpm[wheel_index] = (Ctrl_Data.wheel_freq[wheel_index] * 60) / NUM_TEETH;
 	}
```
Comment (Nomalah) [inline Core/Src/control.c:105]: Can you do a quick check of the formatting?

**Review by Nomalah — CHANGES_REQUESTED** [review]
Undo the formatting changes made for BSPC, and add documentation on the wheel speed sensors in the readme (include the datasheet and the mapping from wss to stm32 pin/timer).

**Comment by Nomalah** [comment]
If you can add the equations used to reach the correct prescalers & auto reload value for each timer into the readme as LaTeX equations that would be great.

**Comment by riyan-noaman** [comment]
merge conflicts need to be resolved again incorrect resolutions were used the first time

**Review by Nomalah — APPROVED** [review]
LGTM

**Pull request merged**